routing.go

```diff
@@ -93,39 +97,55 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key string) (_ []byte, err err
 	ctx, cancel := context.WithTimeout(ctx, time.Minute)
 	defer cancel()
 
-	vals, err := dht.GetValues(ctx, key, 16)
-	if err != nil {
+	var cfg ropts.Options
+	if err := cfg.Apply(opts...); err != nil {
 		return nil, err
 	}
 
-	recs := make([][]byte, 0, len(vals))
-	for _, v := range vals {
-		if v.Val != nil {
-			recs = append(recs, v.Val)
-		}
-	}
-	if len(recs) == 0 {
-		return nil, routing.ErrNotFound
-	}
-
-	i, err := dht.Selector.BestRecord(key, recs)
+	results, err := dht.getValues(ctx, key, &cfg)
 	if err != nil {
 		return nil, err
 	}
 
-	best := recs[i]
-	log.Debugf("GetValue %v %v", key, best)
-	if best == nil {
-		log.Errorf("GetValue yielded correct record with nil value.")
-		return nil, routing.ErrNotFound
+	var outdatedPeers, currentPeers []peer.ID
+
+	var best []byte
+	for result := range results {
+		switch {
+		case result.val == nil:
+			outdatedPeers = append(outdatedPeers, result.from)
+		case best == nil:
+			best = result.val
+			fallthrough
+		case bytes.Equal(result.val, best):
+			currentPeers = append(currentPeers, result.from)
+		default:
+			i, err := dht.Selector.BestRecord(key, [][]byte{best, result.val})
+			if err != nil {
+				log.Error(err)
+				return nil, err
+			}
+			switch i {
+			case 0:
+				outdatedPeers = append(outdatedPeers, result.from)
+			case 1:
+				outdatedPeers = append(outdatedPeers, currentPeers...)
+				currentPeers = append(currentPeers[:0], result.from)
+			default:
+				err := fmt.Errorf("invalid bad selector for key: %s", loggableKey(key))
+				log.Error(err)
+				return nil, err
+			}
+		}
 	}
 
-	fixupRec := record.MakePutRecord(key, best)
-	for _, v := range vals {
-		// if someone sent us a different 'less-valid' record, lets correct them
-		if !bytes.Equal(v.Val, best) {
-			go func(v routing.RecvdVal) {
-				if v.From == dht.self {
+	// if someone sent us a different 'less-valid' record, lets correct them
+	if best != nil && len(outdatedPeers) > 0 {
+		fixupRec := record.MakePutRecord(key, best)
+		for _, p := range outdatedPeers {
+			// TODO: Use a worker.
```

Comment:
This isn't a regression, but we really should have a "corrections" queue somewhere along with a worker pool.

Reply:
Yeah, it would also nice to just have passive repair functionality too. Like, occasionally scan through our records and push them to peers that we have open connections to who are closer than us to the record